yarn-project/world-state/src/native/native_bench.test.ts

```diff
@@ -0,0 +1,292 @@
+      }
+    } else if (insertionType === InsertionType.BATCH) {
+      for (const leafSet of leaves) {
+        await fork.batchInsert(treeId as IndexedTreeId, leafSet as Buffer[], 0);
+      }
+    } else {
+      for (const leafSet of leaves) {
+        await fork.sequentialInsert(treeId as IndexedTreeId, leafSet as Buffer[]);
+      }
+    }
+    const endTime = performance.now();
+    const avgTime = (endTime - startTime) / numBlocks;
+    return avgTime;
+  };
+
+  const runDataRetrievalByIndex = async (
+    numRetrievals: number,
+    treeId: MerkleTreeId,
+    worldState: NativeWorldStateService,
+    op: (treeId: MerkleTreeId, index: bigint, fork: MerkleTreeReadOperations) => Promise<any>,
+  ) => {
+    const fork = await worldState.fork();
+    const treeInfo = await fork.getTreeInfo(treeId);
+    const indices = Array.from({ length: numRetrievals }, () =>
+      Math.floor(Math.random() * Number(treeInfo.size - 1n)),
+    ).map(m => BigInt(m));
+
+    const startTime = performance.now();
+
+    for (const index of indices) {
+      await op(treeId, index, fork);
+    }
+
+    const endTime = performance.now();
+    const avgTime = (endTime - startTime) / numRetrievals;
+    return avgTime;
+  };
+
+  const runDataRetrievalByValue = async (
+    numRetrievals: number,
+    treeId: MerkleTreeId,
+    worldState: NativeWorldStateService,
+    values: (Buffer | Fr)[],
+    op: (treeId: MerkleTreeId, value: Buffer | Fr, fork: MerkleTreeReadOperations) => Promise<any>,
+  ) => {
+    const fork = await worldState.fork();
+    const treeInfo = await fork.getTreeInfo(treeId);
+    const indices = Array.from({ length: numRetrievals }, () =>
+      Math.floor(Math.random() * Number(treeInfo.size - 1n)),
+    ).map(m => BigInt(m));
+
+    const startTime = performance.now();
+
+    for (const index of indices) {
+      const value = values[Number(index)];
+      await op(treeId, value, fork);
+    }
+
+    const endTime = performance.now();
+    const avgTime = (endTime - startTime) / numRetrievals;
+    return avgTime;
+  };
+
+  it.each([
+    [36, 1, 2],
+    [36, 8, 2],
+    [36, 64, 2],
+    [360, 8, 1],
+  ])('Syncs a %s tx block where txs have %s leaves', async (numTxs: number, numLeaves: number, numBlocks: number) => {
+    // We insert a block so we don't obtain any 'first block' advantage
+    await runBlockSyncTest(1, 4, 1, worldState);
+
+    // Now run the test
+    const duration = await runBlockSyncTest(numBlocks, numTxs, numLeaves, worldState);
+    metrics.addBlockSyncMetric(numTxs, numLeaves, duration);
+  });
+
+  it.each([
+    [MerkleTreeId.PUBLIC_DATA_TREE, InsertionType.BATCH, 1],
+    [MerkleTreeId.PUBLIC_DATA_TREE, InsertionType.BATCH, 8],
+    [MerkleTreeId.PUBLIC_DATA_TREE, InsertionType.BATCH, 64],
+    [MerkleTreeId.PUBLIC_DATA_TREE, InsertionType.SEQUENTIAL, 1],
+    [MerkleTreeId.PUBLIC_DATA_TREE, InsertionType.SEQUENTIAL, 8],
+    [MerkleTreeId.PUBLIC_DATA_TREE, InsertionType.SEQUENTIAL, 64],
```

Comment:
I think it would be useful to have a test where the pool of possible slots is small. With `Fr.random()` for slots it's highly unlikely we'd get two txs updating the same slot (e.g. like if both were using the sponsored FPC)